src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,742 @@
+  auto size_recv = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  CHECK_NCCL(ncclGroupStart());
+  for (size_t r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(size_send.ptr(r), 1, ncclUint64, r, *comm, stream));
+    CHECK_NCCL(ncclRecv(size_recv.ptr(r), 1, ncclUint64, r, *comm, stream));
+  }
+  CHECK_NCCL(ncclGroupEnd());
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // allocate merge targets, data transfer...
+  std::vector<SortPiece<VAL>> merge_buffers(num_ranks);
+  std::vector<size_t> aligned_pos_vals_send(num_ranks);
+  std::vector<size_t> aligned_pos_idcs_send(num_ranks);
+
+  size_t buf_size_send_vals_total = 0;
+  size_t buf_size_send_idcs_total = 0;
+  for (size_t i = 0; i < num_ranks; ++i) {
+    // align buffer to allow data transfer of 16byte blocks
+    auto buf_size_vals_recv = get_16b_aligned_count(size_recv[i], sizeof(VAL));
+    merge_buffers[i].values = create_buffer<VAL>(buf_size_vals_recv, Memory::GPU_FB_MEM);
+    merge_buffers[i].size   = size_recv[i];
+
+    aligned_pos_vals_send[i] = buf_size_send_vals_total;
+    buf_size_send_vals_total += get_16b_aligned_count(size_send[i], sizeof(VAL));
+
+    if (argsort) {
+      auto buf_size_idcs_recv  = get_16b_aligned_count(size_recv[i], sizeof(int64_t));
+      merge_buffers[i].indices = create_buffer<int64_t>(buf_size_idcs_recv, Memory::GPU_FB_MEM);
+      aligned_pos_idcs_send[i] = buf_size_send_idcs_total;
+      buf_size_send_idcs_total += get_16b_aligned_count(size_send[i], sizeof(int64_t));
+    } else {
+      merge_buffers[i].indices = create_buffer<int64_t>(0, Memory::GPU_FB_MEM);
+    }
+  }
+
+  // copy values into aligned send buffer
+  auto val_send_buf = local_sorted.values;
+  if (buf_size_send_vals_total > volume) {
+    val_send_buf = create_buffer<VAL>(buf_size_send_vals_total, Memory::GPU_FB_MEM);
+    size_t pos   = 0;
+    for (size_t r = 0; r < num_ranks; ++r) {
+      CHECK_CUDA(cudaMemcpyAsync(val_send_buf.ptr(aligned_pos_vals_send[r]),
+                                 local_sorted.values.ptr(pos),
+                                 sizeof(VAL) * size_send[r],
+                                 cudaMemcpyDeviceToDevice,
+                                 stream));
+      pos += size_send[r];
+    }
+  }
+
+  // copy indices into aligned send buffer
+  auto idc_send_buf = local_sorted.indices;
+  if (argsort && buf_size_send_idcs_total > volume) {
+    idc_send_buf = create_buffer<int64_t>(buf_size_send_idcs_total, Memory::GPU_FB_MEM);
+    size_t pos   = 0;
+    for (size_t r = 0; r < num_ranks; ++r) {
+      CHECK_CUDA(cudaMemcpyAsync(idc_send_buf.ptr(aligned_pos_idcs_send[r]),
+                                 local_sorted.indices.ptr(pos),
+                                 sizeof(int64_t) * size_send[r],
+                                 cudaMemcpyDeviceToDevice,
+                                 stream));
+      pos += size_send[r];
+    }
+  }
+
```

Comment:
If you made copies of the send data, I assume you could destroy it here, as it wouldn't be used later on.

Reply:
Yes,  i will commit my changes regarding cleanup.